db-engine/src/main/java/com/splicemachine/db/iapi/stats/ColumnStatisticsImpl.java

```diff
@@ -228,10 +233,41 @@ public long selectivity(DataValueDescriptor element) {
         // Use Null Data
         if (element == null || element.isNull())
             return nullCount;
+
         // Frequent Items Sketch?
-        long count = frequenciesSketch.getEstimate(element);
-        if (count>0)
-            return count;
+
+        // When look up predicate value in quantilesSketch, the element's data type
+        // must match with the data type stored the ItemsSketch.hashMap
+        // Otherwise the same value of element in ItemsSketch.hashMap can't be queried out
+        // This handle the situation of string predict compare with DATE/TIME/TIMESTAMP
+        DataValueDescriptor lookUpElement = element;
+        int typeFormatId = dvd.getTypeFormatId();
+        int eTypeFormatId = element.getTypeFormatId();
+        boolean isConverted = false;
+        if (typeFormatId != eTypeFormatId) {
+            if (eTypeFormatId == StoredFormatIds.SQL_CHAR_ID) {
+                try {
+                    switch (typeFormatId) {
+                        case StoredFormatIds.SQL_DATE_ID:
+                            lookUpElement = new SQLDate(element.getDate(new GregorianCalendar()));
+                            break;
+                        case StoredFormatIds.SQL_TIME_ID:
+                            lookUpElement = new SQLTime(element.getTime(null));
+                            break;
+                        case StoredFormatIds.SQL_TIMESTAMP_ID:
+                            lookUpElement = new SQLTimestamp(element.getTimestamp(new GregorianCalendar()));
+                    }
+                    isConverted = true;
+                } catch (StandardException e) {
+                    LOG.warn("Data type conversion failure when looking up frequencies.ItemsSketch", e);
+                }
+            }
+        }
+        if (typeFormatId == eTypeFormatId || isConverted) {
```

Comment:
The check of "typeFormatId == eTypeFormatId" may be too strict. Would this affect the number data types? for example, between INT and BIGINT. 
Prefer to restrict the fix to just Char and VarChar type. Something like 
if (eTypeformatId !=StoredFormatIds.SQL_CHAR_ID && eTypeFormatId !=StoredFormatIds.SQL_VARCHAR_ID || isConverted)